### ISSUE TITLE
Confused about SelfSupervisedLoss and labeling with dataset of pairs

### ISSUE THREAD

**Issue by irowberryFS** [KevinMusgrave/pytorch-metric-learning#711]
I have a dataset where each row is a pair of objects (in this case small graphs) and a label indicating if the two objects represent the same thing. A `1` indicates the objects represent the same thing, a `0` indicating that they're different. 
Would it makes sense to use the `SelfSupervisedLoss` wrapper here? I'm guessing no, because I have pairs that aren't augmentations of each other (the ones with label `0`). I suppose I would like justification of what I'm currently doing.

This is what I have. I am using NTXentLoss.

```python
  out1 = self._run_forward(batch["graph1"])
  out2 = self._run_forward(batch["graph2"])

  y = batch["label"].to(bool)
  out = torch.cat((out1, out2))
  loss_fn_labels = torch.arange(len(out), device=out.device)
  labels_2n = torch.cat((y, y))

  temp = loss_fn_labels[labels_2n]
  temp[len(temp) // 2 :] = temp[: len(temp) // 2]

  loss_fn_labels[labels_2n] = temp
  loss = self.loss_fn(out, loss_fn_labels)
  ```

**Comment by KevinMusgrave** [comment]
You can use `SelfSupervisedLoss` if:
- you have two sets of embeddings `x` and `y`
- `x[i]` is a positive pair of `y[i]`
- `x[i]` is a negative pair of `y[j]` for all `j != i`

In your case, I guess it depends if any element in your batch appears more than once, and also if you want to assume all pairs are negative if they aren't explicitly identified as positive. Here's a related discussion: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/705

**Comment by irowberryFS** [comment]
Maybe I'm using the wrong loss?
The only thing that is guaranteed is that if `x[0], y[0]` is labeled `1`, they represent they represent the same thing. If `x[1], y[1]` are also labeled `1`, there is no relation to `x[0], y[0]`. Any pair labeled `0` has nothing in common between the pair, nor anything in common with any positive pairs, nor any of the negatives. 
So to use `SelfSupervisedLoss` I would need arrange my pairs so that `x` is `n` positive pairs long. Then `y[:n]` is the matching part of `x`, and `y[n:]` is everything else?
Something like this?
```python
  out1 = self._run_forward(batch["graph1"]) # one side of pair
  out2 = self._run_forward(batch["graph2"]) # other side
  label = batch['label']
  
  x = out1[label == 1]
  y = out2[label == 1]
  y = torch.cat((y, out1[label==0], out2[label==0]))
```

**Comment by KevinMusgrave** [comment]
Do you want `x[0]` and `x[1]` to be considered a negative pair? What about `x[0]` and `y[1]`?

If the answer to both of these is "no" then you can't use SelfSupervisedLoss and you also can't use NTXentLoss (see this explanation: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/705#issuecomment-2249299897)

**Comment by irowberryFS** [comment]
Yes that's correct, both of those are negative. Only pairs with label 1  in the dataset are positive pairs, every other combination can be a negative pair.

**Comment by KevinMusgrave** [comment]
I was about to say that you can use SelfSupervisedLoss, but I forgot you have pairs that you explicitly mark as negative, so you still can't use SelfSupervisedLoss since it assumes `x[i] y[i]` are always a positive pair.

So you'll have to use NTXentLoss instead. I'm not 100% sure what your code does in your opening post, but basically you need to make the labels so that if two samples have the same label they are considered a positive. Otherwise they will be considered a negative.

**Comment by irowberryFS** [comment]
Okay, I that's what my hacky slicing does. Thanks for the quick replies! Much appreciated.
